diff --git a/extopenscad.hs b/extopenscad.hs
index 7ae05fe..281669a 100644
--- a/extopenscad.hs
+++ b/extopenscad.hs
@@ -23,7 +23,7 @@ import Data.AffineSpace
 import Control.Applicative
 -- The following is needed to ensure backwards/forwards compatibility
 -- make sure we don't import (<>) in new versions.
-import Options.Applicative (fullDesc, progDesc, header, auto, info, helper, help, str, argument, switch, value, long, short, option, metavar, nullOption, reader, execParser, Parser)
+import Options.Applicative (fullDesc, progDesc, header, auto, info, helper, help, str, argument, switch, value, long, short, option, metavar, execParser, Parser)
 import System.FilePath
 
 -- Backwards compatibility with old versions of Data.Monoid:
@@ -58,8 +58,10 @@ formatExtensions =
 	, ("obj", OBJ)
 	]
 
-readOutputFormat :: String -> Maybe OutputFormat
-readOutputFormat ext = lookup (map toLower ext) formatExtensions
+readOutputFormat :: Monad m => String -> m (OutputFormat)
+readOutputFormat ext = case lookup (map toLower ext) formatExtensions of
+	Nothing -> fail ("unknown extension: "++ext)
+	Just x -> return x
 
 guessOutputFormat :: FilePath -> OutputFormat
 guessOutputFormat fileName =
@@ -71,28 +73,25 @@ guessOutputFormat fileName =
 extOpenScadOpts :: Parser ExtOpenScadOpts
 extOpenScadOpts =
 	ExtOpenScadOpts
-	<$> nullOption
+	<$> option (pure <$> str)
 		(  short 'o'
 		<> long "output"
 		<> value Nothing
 		<> metavar "FILE"
-		<> reader (pure . str)
 		<> help "Output file name"
 		)
-	<*> nullOption
+	<*> option (pure <$> (readOutputFormat =<< str))
 		(  short 'f'
 		<> long "format"
 		<> value Nothing
 		<> metavar "FORMAT"
 		<> help "Output format"
-		<> reader (pure . readOutputFormat)
 		)
-	<*> option
+	<*> option (pure <$> auto)
 		(  short 'r'
 		<> long "resolution"
 		<> value Nothing
 		<> metavar "RES"
-		<> reader (pure . auto)
 		<> help "Approximation quality"
 		)
 	<*> switch
diff --git a/implicit.cabal b/implicit.cabal
index 0b30bde..17026f0 100644
--- a/implicit.cabal
+++ b/implicit.cabal
@@ -20,7 +20,7 @@ Library
         base >= 3 && < 5,
         filepath,
         directory,
-        optparse-applicative >= 0.9.0.0 && < 0.10.0,
+        optparse-applicative >= 0.10.0,
         parsec,
         unordered-containers,
         parallel,
